### PR TITLE
Clean up storage layer

### LINKED ISSUE #344: RF storage backends code to provide uniform Storage interface to switch between implementations
ATM minio-vs-s3 code logic is not centralized and spread across multiple files
```
$> git grep -i -l minio | grep -v -e test | grep '/api.*\.py'
dandiapi/api/_copy.py
dandiapi/api/checksum.py
dandiapi/api/models/upload.py
dandiapi/api/multipart.py
dandiapi/api/storage.py
dandiapi/api/views/asset.py
```
with code switching  based on the storage instance class:
```
$> git grep -i 'isinstance(.*minio' 
dandiapi/api/_copy.py:    elif isinstance(storage, MinioStorage):
dandiapi/api/checksum.py:        if isinstance(storage, MinioStorage):
dandiapi/api/models/upload.py:        elif isinstance(storage, MinioStorage):
dandiapi/api/multipart.py:            if isinstance(storage, MinioStorage):
dandiapi/api/views/asset.py:    elif isinstance(storage, MinioStorage):
```

Having established all currently needed interfaces/functionality from storage, and with the embargo design in mind (#158, https://github.com/dandi/helpdesk/issues/12) I think it would be good time to RF the codebase to provide a consistent interface to the storage backends, to make it also easier (without requiring to place ad-hoc conditions throughout codebase) support other storage backends (wasabi #147, which might be most relevant for embargo due to lower cost AFAIK)

### PR DESCRIPTION
**Pull request by danlamanna**
Fixes #344.

This just moves storage operations to their storage backend instead of if/else switching throughout the rest of the codebase. This should make it easier to reason about the rest of the code, and allow us to see exactly which features we're using of S3/minio at a glance.

### CODE REVIEW

File: dandiapi/api/tests/test_zarr_upload.py [hunk 1]
```diff
@@ -152,7 +152,6 @@ def test_zarr_rest_upload_complete_missing_file(
 
     resp = authenticated_api_client.post(f'/api/zarr/{zarr_archive.zarr_id}/upload/complete/')
     assert resp.status_code == 400
-    assert resp.json() == [f'File {upload.path} does not exist.']
```
Comment (mvandenburgh) [inline dandiapi/api/tests/test_zarr_upload.py:155]: Why is this removed?

**Review by mvandenburgh — COMMENTED** [review]
Changes look correct to me, just one question

**Review by mvandenburgh — APPROVED** [review]

**Comment by danlamanna** [comment]
I want to wait to merge this until after #1220.

File: dandiapi/api/tests/test_zarr_upload.py (hunk 1 above)
Comment (AlmightyYakob) [reply dandiapi/api/tests/test_zarr_upload.py:155]: What was the reason?

**Review by AlmightyYakob — APPROVED** [review]
Looks good!

File: dandiapi/api/tests/test_zarr_upload.py (hunk 1 above)
Comment (mvandenburgh) [reply dandiapi/api/tests/test_zarr_upload.py:155]: I believe it was because the changes in #1220 made it obsolete (@danlamanna can correct me if i'm wrong)

File: dandiapi/api/storage.py [hunk 2]
```diff
@@ -1,13 +1,66 @@
...
+        self.h.update(bytes)
+
+    @property
+    def checksum(self):
+        return self.h.hexdigest()
+
+
+class DandiMultipartMixin:
+    @staticmethod
+    def _iter_part_sizes(file_size: int, part_size: int = None) -> Iterator[tuple[int, int]]:
```
Comment (mvandenburgh) [inline dandiapi/api/storage.py:37]:
```suggestion
    def _iter_part_sizes(file_size: int, part_size: int | None = None) -> Iterator[tuple[int, int]]:
```

Comment (danlamanna) [reply dandiapi/api/storage.py]: I did you one better and removed the parameter. :)
